source/Cosmos.HAL2/SoundBlaster16.cs

```diff
@@ -0,0 +1,334 @@
+﻿using System;
+using Cosmos.Core;
+using Cosmos.HAL.Drivers.PCI.Audio.Generic;
+
+namespace Cosmos.HAL
+{
+    public unsafe class SoundBlaster16 : SoundCard
+    {
+        private bool sound_blaster = false;
+        private byte sb16_version_major;
+        private byte sb16_version_minor;
+        /// <summary>
+        /// Gets weather or not a soundblaster 16 is present on the system
+        /// </summary>
+        public bool IsPresent { get { return sound_blaster; } }
+        /// <summary>
+        /// Gets the DSP Version
+        /// </summary>
+        public Version DSPVersion { get { return new Version(sb16_version_major, sb16_version_minor); } }
+        /// <summary>
+        /// Gets weather or not audio is playing.
+        /// </summary>
+        public bool IsAudioPlaying { get { return PlayingSound; } }
+
+        #region DSP Ports
+        const ushort DSP_RESET = 0x226;
+        const ushort DSP_READ = 0x22A;
+        const ushort DSP_WRITE = 0x22C;
+        const ushort DSP_READSTATUS = 0x22E;
+        #endregion
+
+        #region Mixer Commands
+        const byte MIXER_SETIRQ = 0x80;
+        const byte MIXER_SETVOL = 0x22;
+        #endregion
+
+        #region DSP Commands
+        const byte DSP_EnableSpeaker = 0xD1;
+        const byte DSP_DisableSpeaker = 0xD3;
+        const byte DSP_GETVERSION = 0xE1;
+        const byte DSP_SETTIME = 0x40;
+        const byte DSP_SETPLAYBACKRATE = 0x41;
+        #endregion
+
+        #region Mixer Ports
+        const ushort MixerPort = 0x224;
+        const ushort MixerDataPort = 0x225;
+        #endregion
+        private byte[] Buffer = new byte[64000]; //64K Buffer
+        private byte* bufferinmem;
```

Comment:
Can you please replace the byte* with a `MemoryBlock`? This should make the HAL project again not need unsafe code. As a rule of thumb, no driver in HAL should require using unsafe code.